### PR TITLE
feat: support for rollback in case of error in intermediate steps

### PR DESCRIPTION
**Pull request by 0zyt**
Signed-off-by: 0zyt <zyt0@duck.com>

## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [ ] I have added relevant tests

## Description
When an error occurs while updating, roll back

## Related Issues
https://github.com/devstream-io/devstream/issues/1173

## New Behavior (screenshots if needed)

### CODE REVIEW

**Comment by daniel-hutao** [comment]
@KeHaohaoke Please help to fix the ci error here. It seems like a configuration issue with build machine.

File: internal/pkg/upgrade/upgrade.go [hunk 1]
```diff
@@ -137,30 +137,67 @@ func parseVersion(old, new string) (*semver.Version, *semver.Version, error) {
...
+				}
+
+			//the error occur in the step 3
+			case 2:
+				if err := os.Rename(dtmFilePath, dtmTmpFilePath); err != nil {
+					log.Debugf("Dtm upgrade rollback error: %s", err.Error())
+				}
+
+			//the error occur in the step 4
+			case 3:
```
Comment (daniel-hutao) [inline internal/pkg/upgrade/upgrade.go:168]: If we failed in step 4, it seems like step 321 all should be rollback.

1 failed -> rollback 1
2 failed -> rollback 2,1
3 failed -> rollback 3,2,1
4 failed -> rollback 4,3,2,1

**Review by daniel-hutao — CHANGES_REQUESTED** [review]

**Review by aFlyBird0 — COMMENTED** [review]
It looks like your code has been off the main branch for two months.

Please run `git fetch upstream && git rebase upstream/main` on the current branch first to update the code.
(If you are not particularly proficient with git, it is recommended that you back up your current code changes first)

@0zyt

**Review by daniel-hutao — APPROVED** [review]
lgtm

**Review by aFlyBird0 — APPROVED** [review]
Good job!

**Pull request merged**